src/Widgets/EndSessionDialog.vala

```diff
@@ -155,6 +155,17 @@ public class Session.Widgets.EndSessionDialog : Gtk.Window {
             destroy ();
         });
 
+        key_press_event.connect ((event) => {
+            // Close dialogs when Escape key is pressed
+            if (event.keyval == 65307) {
```

Comment:
I think there are better ways, like: 
https://github.com/elementary/photos/blob/9022ce00b6041e0e98f6dbe7936f8fd1ee9d5623/src/direct/DirectWindow.vala#L118-L124
https://github.com/elementary/shortcut-overlay/blob/f2e9cdd1d6073cc0265b3c8e574cbea65163308c/src/Application.vala#L40-L61
https://github.com/peteruithoven/resizer/blob/2067abf21ad304f9d52f06e1a02ac02087abfbcb/src/Application.vala#L33-L40

Reply:
I was concerned that this line is not readable but didn't know to make it look even better. Thank you :blush:

Reply:
I've inquired on why one would use an action over a key_press_events signal handler and got a insightful response, you can activate an action also from other places. This would allow you to reuse the logic in the cancel button handler. (By putting that logic in an a action, calling that action from the cancel button handler and with an accelerator (keyboard shortcut). 
Something similar is done in the keyboard shortcut overlay:
https://github.com/elementary/shortcut-overlay/blob/f2e9cdd1d6073cc0265b3c8e574cbea65163308c/src/Application.vala#L40-L61

Reply:
Sounds good, but because `add_action ()` is in `GLib.ActionMap` which `Granite.MessageDialog` does not inherit and `set_accels_for_action ()` methods are only available in `Gtk.Application`, maybe something like this would work?
(I'm sorry if I am misunderstanding…)

```diff
--- a/src/Widgets/EndSessionDialog.vala
+++ b/src/Widgets/EndSessionDialog.vala
@@ -157,9 +157,7 @@ public class Session.Widgets.EndSessionDialog : Gtk.Window {
 
         key_press_event.connect ((event) => {
             if (Gdk.keyval_name (event.keyval) == "Escape") {
-                server.canceled ();
-                server.closed ();
-                destroy ();
+                cancel.clicked ();
             }
 
             return false;
```

Reply:
That would be another thing I didn't know. 
I think it's bad practice to emulate clicks simply to reuse logic. 
So in this case I would make the action and activate it in both the cancel button and key press handler. Or create a separate quit function, but that would require also storing the server as a EndSessionDialog instance property.

Reply:
> So in this case I would make the action and activate it in both the cancel button and key press handler.

I see, it would be like this?

```diff
--- a/src/Widgets/EndSessionDialog.vala
+++ b/src/Widgets/EndSessionDialog.vala
@@ -149,17 +149,20 @@ public class Session.Widgets.EndSessionDialog : Gtk.Window {
 
         cancel.grab_focus ();
 
-        cancel.clicked.connect (() => { 
+        var cancel_action = new SimpleAction ("cancel", null);
+        cancel_action.activate.connect (() => {
             server.canceled ();
             server.closed ();
             destroy ();
         });
 
+        cancel.clicked.connect (() => {
+            cancel_action.activate (null);
+        });
+
         key_press_event.connect ((event) => {
             if (Gdk.keyval_name (event.keyval) == "Escape") {
-                server.canceled ();
-                server.closed ();
-                destroy ();
+                cancel_action.activate (null);
             }
 
             return false;
```

Reply:
Yep. (And just to be sure I wasn't giving bad advice again, I've tested that that works)